text/3502-cargo-script.md

```diff
@@ -0,0 +1,1229 @@
+Users would need to know to run `cargo clean --manifest-path foo.rs` before
+deleting `foo.rs` or to just do `rm -rf ~/.cargo/target` and wipe everything
+(since the specific target directory will be non-obvious).
+
+In the future we can
+[track embedded manifests and garbage collect their `target/`](https://github.com/rust-lang/cargo/issues/12633).
+
+## `cargo new` support
+
+A `cargo new <flag> foo.rs` could generate a source file with an embedded manifest.
+
+Questions
+- Should we do this implicitly with the `.rs` extension?
+- If we have a flag, what should we name it?
+
+## A shorter flag for `--manifest-path`
+
+Cargo users are used to the `--manifest-path` argument being inferred from
+their current working directory but that doesn't work for embedded manifests.
+
+It would be helpful if we had an alias for this argument to make it easier to
+specify, whether it was a short word for a "short flag".
+
+Options include
+- `--script` with `-s` or `-S` for a short flag, but is the meaning clear
+  enough?  What about in the context of multi-file packages taking advantage
+  of it?
+- `p` is taken by `--package`
+- `-m`, `-M`, and `-P` are available, but are the meanings clear enough?
+
+## Cleaner output
+
+`cargo foo.rs` is just a wrapper around `cargo run --manifest-path foo.rs` and cargo can be quite noisy.
+The out-of-tree prototype for this instead ran `cargo run --quiet
+--manifest-path foo.rs` but that was confusing as it is unclear when a long
+execution time is from a slow compile or from the program.
+
+In the future, we could try to find ways to better control cargo's output so at
+least `cargo foo.rs` is less noisy but appears responsive.
+
+See
+[t-cargo zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/Re-thinking.20cargo's.20output)
+and [rust-lang/cargo#8889](https://github.com/rust-lang/cargo/issues/8889).
+
+Note: `--quiet` is inferred when piping to a file ([rust-lang/cargo#12305](https://github.com/rust-lang/cargo/pull/12305))
+
+## Executing `<stdin>`
+
+We could extend this to allow accepting single-file packages from stdin, either
+explicitly with `-` or implicitly when `<stdin>` is not interactive.
+
+## Implicit `main` support
+
+Like with doc-comment examples, we could support an implicit `main`.
+
+Ideally, this would be supported at the language level
+- Ensure a unified experience across the playground, `rustdoc`, and `cargo`
+- `cargo` can directly run files rather than writing to intermediate files
+  - This gets brittle with top-level statements like `extern` (more historical) or bin-level attributes
+
+Behavior can be controlled through editions
+
+## `[lib]` support
+
+In an effort to allow low-overhead packages in a workspace, we may also allow `[lib]`s to be defined.
+
+A single-file package may only be a `[bin]` or a `[lib]` and not both.
+
+We would support depending on these, publishing them, etc.
```

Comment:
publishing them means that the packages would be published without readme and license files and most likely other metadata though, so I'm a bit hesitant regarding this...

Reply:
As this is a future possibility, this is more helpful to add flavor / context for discussions and decisions.  We can work those issues out if/when we come to this.